File: uns/java/app/src/main/java/practicos/practico6/Tree.java
package practicos.practico6;

import java.util.Iterator;

import practicos.practico4.Lista;
import practicos.practico4.interfaces.Position;
import practicos.practico4.interfaces.PositionList;
import practicos.practico6.interfaces.TreeTDA;
import practicos.practico6.exceptions.*;

public class Tree<E> implements TreeTDA<E> {
    private TNode<E> root;
    private int size;

    public Tree() {
        this.root = null;
        this.size = 0;
    }

    @Override
    public int size() {
        return this.size;
    }

    @Override
    public boolean isEmpty() {
        return this.size == 0;
    }

    @Override
    public Iterator<E> iterator() {
        return new TreeIterator<>(this.root);
    }

    @Override
    public Iterable<Position<E>> positions() {
        PositionList<TNode<E>> positions = new Lista<TNode<E>>();
        
        Iterator<E> currentTreeIterator = new TreeIterator<>(this.root);

        while (currentTreeIterator.hasNext()) {
            positions.addFirst(new TNode<E>(currentTreeIterator.next()));
        }

        return positions;
    }

    @Override
    public E replace(Position<E> v, E e) throws InvalidPositionException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'replace'");
    }

    @Override
    public Position<E> root() throws EmptyTreeException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'root'");
    }

    @Override
    public Position<E> parent(Position<E> v) throws InvalidPositionException, BoundaryViolationException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'parent'");
    }

    @Override
    public Iterable<Position<E>> children(Position<E> v) throws InvalidPositionException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'children'");
    }

    @Override
    public boolean isInternal(Position<E> v) throws InvalidPositionException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'isInternal'");
    }

    @Override
    public boolean isExternal(Position<E> v) throws InvalidPositionException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'isExternal'");
    }

    @Override
    public boolean isRoot(Position<E> v) throws InvalidPositionException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'isRoot'");
    }

    @Override
    public void createRoot(E e) throws InvalidOperationException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'createRoot'");
    }

    @Override
    public Position<E> addFirstChild(Position<E> p, E e) throws InvalidPositionException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'addFirstChild'");
    }

    @Override
    public Position<E> addLastChild(Position<E> p, E e) throws InvalidPositionException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'addLastChild'");
    }

    @Override
    public Position<E> addBefore(Position<E> p, Position<E> rb, E e) throws InvalidPositionException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'addBefore'");
    }

    @Override
    public Position<E> addAfter(Position<E> p, Position<E> lb, E e) throws InvalidPositionException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'addAfter'");
    }

    @Override
    public void removeExternalNode(Position<E> p) throws InvalidPositionException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'removeExternalNode'");
    }

    @Override
    public void removeInternalNode(Position<E> p) throws InvalidPositionException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'removeInternalNode'");
    }

    @Override
    public void removeNode(Position<E> p) throws InvalidPositionException {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'removeNode'");
    }
    
}

File: uns/rust/src/implementations/Tree.rs
// Linked List
use crate::implementations::LinkedList::LinkedList;
use crate::interfaces::LinkedListInterface::LinkedListInterface;
// Tree
use crate::interfaces::TreeInterface::TreeInterface;
// TreeNode
use crate::implementations::TreeNode::TreeNode;
use crate::interfaces::TreeNodeInterface::TreeNodeInterface;

#[derive(Copy, Clone, Debug)]
pub struct Tree<T: PartialEq + Copy> {
root: Option<*mut TreeNode<T>>,
len: usize,
}

impl<T: PartialEq + Copy> TreeInterface<T> for Tree<T> {
    // Constructor
    fn new() -> Self {
        Tree { 
            root: None,
            len: 0,
        }
    }

    fn new_with_root(root_node: *mut TreeNode<T>) -> Self {
        Tree {
            root: Some(root_node),
            len: 1,
        }
    }

    // booleans
    fn is_empty(&self) -> bool {
        self.len == 0
    }

    fn is_internal(&self, node: *mut TreeNode<T>) -> bool {
        let mut result = false;
        if self.root.is_some() {
            for pos in self.positions().iterator() {
                unsafe {
                    if pos == *node {
                        if (*pos.get_children()).get_size() == 0 {
                            result = true;
                        }
                    }
                }
            }
        }
        result
    }

    fn is_external(&self, node: *mut TreeNode<T>) -> bool {
        let mut result = true;
        if self.root.is_some() {
            for pos in self.positions().iterator() {
                unsafe {
                    if pos == *node {
                        if (*pos.get_children()).get_size() == 0 {
                            result = false;
                        }
                    }
                }
            }
        }
        result
    }

    fn is_root(&self, node: *mut TreeNode<T>) -> bool {
        let mut result = false;
        if self.root.is_some() {
            unsafe {
                if *self.root.unwrap() == *node {
                    result = true; 
                }
            }
        }
        result
    }

    // getters
    fn get_root(&self) -> Option<*mut TreeNode<T>> {
        self.root
    }

    fn get_all_elements(&self) -> LinkedList<&T> {
        let mut remaining_nodes: LinkedList<TreeNode<T>> = LinkedList::new();
        let mut result_list: LinkedList<&T> = LinkedList::new();
        let mut current_node = self.root;

        unsafe {
            while let Some(node) = current_node {
                // First, we push the current node value into our result list.
                result_list.push_back((*node).get_value());
                // Then we get an iterator for the node's children list. 
                let children = (*(*node).get_children()).iterator();
                
                // Pushing all the node's children into the remaining nodes list.
                for node in children {
                    remaining_nodes.push_back(node);
                }

                // updating current_node to be the next in the remaining list.
                if remaining_nodes.get_size() > 0 {
                    current_node = Some(Box::into_raw(Box::new(*remaining_nodes.pop_front().unwrap())));
                } else {
                    current_node = None;
                }
            }
        }
        result_list
    }

    fn insert_node(&mut self, node: &mut TreeNode<T>) {
        // We'll insert the node in the root of the three.
        match self.root {
            Some(old_root) => {
                unsafe {
                    node.push_child_front(*old_root);
                    self.root = Some(Box::into_raw(Box::new(*node)));
                    self.len += 1;
                }
            }
            None => {
                self.root = Some(Box::into_raw(Box::new(*node)));
                self.len += 1;
            }
        }
    }

    fn positions(&self) -> LinkedList<TreeNode<T>> {
        let mut result_list: LinkedList<TreeNode<T>> = LinkedList::new();
        if self.get_root().is_some() {
            unsafe {
                result_list.push_back(*self.get_root().unwrap());
                let node_children = *(*self.get_root().unwrap()).get_children();

                if node_children.get_size() > 0 {
                    for mut pos in node_children.iterator() {
                        self.child_positions(&mut pos, &mut result_list);
                    }
                }
            }
        }
        result_list
    }




    fn replace(&self, position: TreeNode<T>, new_node: TreeNode<T>) {
        // To replace a node, we're going to remove it from the parent's children list. and add the
        // new node in the same position of the list.
        if self.root.is_some() {
            unsafe {
                for node in self.positions().iterator() {
                    for child in (*node.get_children()).iterator() {
                        if child == position {
                            (*node.get_children()).push_front(new_node);
                            return;
                            }
                        } 
                    }
                }
        }
    }

    // setters
    fn set_root(&mut self, root_value: T) {
        if self.root.is_none() {
            self.len += 1;
        }

        let new_root = Some(Box::into_raw(Box::new(TreeNode::new(root_value))));
        self.root = new_root;
    }

    fn add_first_child(&mut self, position: TreeNode<T>, value: T) {
        if self.root.is_some() {
            let list_of_nodes = self.positions();
            let new_child = TreeNode::new(value);

            for mut node in list_of_nodes.iterator() {
                if position == node {
                    node.push_child_front(new_child);
                    break;
                }
            }
            self.len += 1;
        }
    }

    fn add_last_child(&mut self, position: TreeNode<T>, value: T) {
        if self.root.is_some() {
            let list_of_nodes = self.positions();
            let new_child = TreeNode::new(value);

            for mut node in list_of_nodes.iterator() {
                if position == node {
                    node.push_child_back(new_child);
                    break;
                }
            }
            self.len += 1;
        }
    }

    fn add_before(&mut self, position: TreeNode<T>, value: T) {
        if self.root.is_some() {
            let list_of_nodes = self.positions();
            let new_child = TreeNode::new(value);

            for mut node in list_of_nodes.iterator() {
                if position.get_value() == node.get_value() {
                    node.push_child_front(new_child);
                    break;
                }
            }
        }
    }

    fn add_after(&mut self, position: TreeNode<T>, value: T) {
        if self.root.is_some() {
            let list_of_nodes = self.positions();
            let new_child = TreeNode::new(value);

            for mut node in list_of_nodes.iterator() {
                if position.get_value() == node.get_value() {
                    node.push_child_back(new_child);
                    break;
                }
            }
        }
    }
}

// Method developed to add the remaining nodes inside the list passed by parameter.
impl<T: Copy + PartialEq> Tree<T> {
    fn child_positions(&self, new_root: *mut TreeNode<T>, list_of_nodes: *mut LinkedList<TreeNode<T>>)  {
        unsafe {
            let node_children = *(*new_root).get_children();
            (*list_of_nodes).push_back(*new_root);
            if node_children.get_size() > 0 {
                for mut child in node_children.iterator() {
                    self.child_positions(&mut child, list_of_nodes);
                }
            }
        }
    } 
}
